### DISCUSSION TITLE
Global variables and program state

### DISCUSSION THREAD

**Discussion by jojogib** [RomanKubiak/ctrlr#582] [Q&A]
Hey folks,

I almost finished a panel but I have a major problem.
I have 16 channels each channel contains 100 programs which you could edit separately
I solved it by adding 2 slider with global varibles one for channel k0 (1-16) and one for program k1 (1-100)

This works fine but is there a simple way to keep track of values. When I change the global variable sliders from channel 1 program 1 to channel 14 program 58 so that I can see the actual state of the modulators?

Thank you for your help

**Comment by dnaldoog** [comment]
Not   sure if I understand exactly what you are after, but if you mean that when you select channel 14, program 58 the sliders remember the values for that setting and then change to another value when you change the channel and program to a completely different setting then I would probably do that in lua with an array matrix
<pre>
t={
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, -- 01
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, , --02
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, , --03
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, --04
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, , --05
--
{0,0,0,0,0,0,0,0,0,0,0,0,0,33,0,0},  --58
--
{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, --100
}

local channel=panel:getGlobalVariable(0)
local program=panel:getGlobalVariable(1)

panel:getModulatorByName("someSliderOrCombo"):setModulatorValue(t[program][channel],false,false,false) -- set to 33
</pre>
Another way would be to give each modulator a sequential `modulatorCustomIndex` number and write data to a single serialised table or MemoryBlock, for setting and getting.

If you are really stuck even after this explanation, please post the panel and I can try and help you work out what is the best solution.

**Comment by jojogib** [comment]
Dnaldoog thank you for your answer!!
Here is a simplification. Depending on the value of glob 0 and glob 1 I wanna see the correct value of reso and cutoff.
[glob var2.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/13548070/glob.var2.bpanelz.zip)

**Comment by dnaldoog** [comment]
That was difficult!

Try this! - it appears that global variables `k0`,`k1` etc don't seem to work in recording values, so I created two global lua variables `pr` and `ch` for patch(patch) / channel. It seems to be working. Please let me know.

[filter array matrix storage system.zip](https://github.com/RomanKubiak/ctrlr/files/13555930/filter.array.matrix.storage.system.zip)

----

<pre>
constructor = function()
    pr = 0 -- program or patch 1~100 (0~99)
    ch = 0 -- channel 1-16 (0~15)
    myData = {} -- initialize
    for i = 0, 99 do -- initialize and array of 100 arrays within myData
        myData[i] = {}
     -- initialize
    end

    for i = 0, 99 do -- in each of the 100 arrays assign 16 * 2 pairs of data values
        for j = 0, 15 do
            myData[i][j] = {reso = 0, cutoff = 0}
         -- initialize
        end
    end

    -- end create array matrix of values
end


setFilterValues = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop from running at boot
    if source ~= 4 then
        return
    end
    local name = L(mod:getName())
    local t = {reso = true, cutoff = true}
    for k in pairs(t) do
        myData[pr][ch][k] = panel:getComponent(k):getValue()
    end
end

getFilterValues = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop from running at boot
    if source ~= 4 then
        return
    end
    local name = L(mod:getName())

    if name == "channel" then
        ch = value
    elseif name == "patch" then
        pr = value
    end

    panel:getComponent("reso"):setValue(myData[pr][ch].reso, true)
    panel:getComponent("cutoff"):setValue(myData[pr][ch].cutoff, true)
end
</pre>

**Comment by dnaldoog** [comment]
Here's another panel that uses one big 3200 byte `MemoryBlock()`!

***Edit**: Version 2.0 was not working. Version 3.0 should work and this panel saves state.

[array matrix storage system_3_0.zip](https://github.com/RomanKubiak/ctrlr/files/13575792/array.matrix.storage.system_3_0.zip)

**Comment by jojogib** [comment]
Hey many thanks!
I had sucsess with your first panel. Are there any problems if I do 70+ sliders with this method instead of only this 2?
And I do have another question. There is this demo panel called wavform conversion but i always get errror when I try to use it. 
[DEMO - Waveform conversion.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/13615805/DEMO.-.Waveform.conversion.bpanelz.zip)
Is it possible to load a wav and dump it via sysex (found this documantation http://midi.teragonaudio.com/tech/sds.htm). For the first step I would be happy if I could load a wav-file and convert it into the HEX i can dump afterwards. Here is a blank panel with a small wav
[dumpwav.zip](https://github.com/RomanKubiak/ctrlr/files/13615700/dumpwav.zip)
Thank you very much

**Comment by dnaldoog** [comment]
You should be able to scale up that panel with this code:
[array matrix storage system_4_0_table.zip](https://github.com/RomanKubiak/ctrlr/files/13623091/array.matrix.storage.system_4_0_table.zip)

<pre>
--
-- Called when the panel is created, no modulators will exist at this point
-- consider this the panels constructor
--
constructor = function()
    pr = 0 -- program or patch 1~100 (0~99)
    ch = 0 -- channel 1-`16 (0~15)

    myData = {} -- initialize
    for i = 0, 99 do -- initialize and array of 100 arrays within myData
        myData[i] = {}
        -- initialize
    end

    for i = 0, 99 do -- in each of the 100 arrays assign 16 * 2 pairs of data values
        for j = 0, 15 do
            for k in pairs(set) do
                myData[i][j] = {}
                myData[i][j][k] = 0
                -- initialize
            end
        end
    end

    -- end create array matrix of values
end

set = {
    reso = true,
    cutoff = true,
    ["lfo rate"] = true,
    ["lfo delay"] = true
}

getFilterValues = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop from running at boot
    if source ~= 4 then
        return
    end
    local name = L(mod:getName())

    if name == "channel" then
        ch = value
    elseif name == "patch" then
        pr = value
    end
    for k in pairs(set) do
        panel:getComponent(k):setValue(myData[pr][ch][k] or 0, true)
    end
end

setFilterValues = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop from running at boot
    if source ~= 4 then
        return
    end
    local name = L(mod:getName())
    myData[pr][ch][name] = value
end

</pre>

----
I tried adding your Sine Wave to Atom's panel but you are right; it is crashing looking for a function `int64ToInt`.
I tried `math.int64ToInt` but no good.

You might need to refer to [this page](https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/Lua/JuceClasses/LAudio.cpp#L83) but I don't know much about it I'm afraid and that panel of Atom's is beyond me.
